services/src/test/java/org/apache/druid/testing/embedded/EmbeddedDruidCluster.java

```diff
@@ -231,13 +242,50 @@ public TestFolder getTestFolder()
   }
 
   /**
-   * The embedded Zookeeper server used by this cluster, if any.
-   *
-   * @throws NullPointerException if this cluster has no embedded zookeeper.
+   * Hostname to be used for embedded services (both Druid or external).
+   * Using this hostname ensures that the underlying service is reachable by both
+   * EmbeddedDruidServers and DruidContainers.
+   */
+  public String getEmbeddedServiceHostname()
+  {
+    return hasDruidContainers ? getDefaultHost() : "localhost";
+  }
+
+  /**
+   * Hostname for the host machine running the containers. Using this hostname
+   * instead of "localhost" allows all the Druid containers to talk to each
+   * other and also other EmbeddedDruidServers.
    */
-  public EmbeddedZookeeper getZookeeper()
+  public static String getDefaultHost()
   {
-    return Objects.requireNonNull(zookeeper, "No embedded zookeeper configured for this cluster");
+    try {
+      return InetAddress.getLocalHost().getHostAddress();
+    }
+    catch (UnknownHostException e) {
+      throw new ISE(e, "Unable to determine host name");
+    }
+  }
+
+  /**
+   * Replaces {@code localhost} or {@code 127.0.0.1} in the given connectUri
+   * with {@link #getEmbeddedServiceHostname()}. Using the embedded URI ensures
+   * that the underlying service is reachable by both EmbeddedDruidServers and
+   * DruidContainers.
+   */
+  public String getEmbeddedConnectUri(String connectUri)
+  {
+    if (!hasDruidContainers) {
+      return connectUri;
+    } else if (connectUri.contains("localhost")) {
+      return StringUtils.replace(connectUri, "localhost", getEmbeddedServiceHostname());
```

Comment:
Seeing URI manipulations like this stresses me out ;P

It's just test code, but it's still better form to manipulate the URL properly. That means parsing it with `new URI(connectUri)` and then creating a replacement `URI` using the 7-arg constructor.